### ISSUE TITLE
Cannot open settings/keymap files from SSH remote workspaces

### ISSUE THREAD

**Issue by brucegorman** [zed-industries/zed#55554]
### Reproduction steps

1. Start Zed
2. Press `Ctrl+Alt+,` (or run `zed: open settings file` from the command palette).

### Current vs. Expected behavior

The settings file, while it is in the correct place and is not a symlink, is not opened. Nothing happens.

### Zed version and system specs

Zed: v1.0.0+stable.257.5ec84a926ef83865afb92d2a3d1ca3b419572cf9 (Zed) 
OS: Windows 10.0.26200
Memory: 31.6 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics || Intel Corporation || 32.0.101.6790

### Attach Zed log file

<details><summary>Zed.log</summary>

```log
2026-05-03T22:06:09+08:00 INFO  [remote::transport::ssh] Remote shell discovered: /bin/bash
2026-05-03T22:06:11+08:00 INFO  [remote::transport::ssh] Remote platform discovered: RemotePlatform { os: Linux, arch: X86_64 }
2026-05-03T22:06:14+08:00 WARN  [remote_server] (remote server) stdin read failed: unexpected end of file
2026-05-03T22:06:14+08:00 WARN  [remote_server] (remote server) error reading message on stdin, dropping connection.
2026-05-03T22:06:14+08:00 INFO  [remote_server] (remote server) accepting new connections
2026-05-03T22:06:14+08:00 INFO  [remote_server] (remote server) accepted new connections
2026-05-03T22:06:14+08:00 WARN  [remote_server] (remote server) stdin read failed: unexpected end of file
2026-05-03T22:06:14+08:00 WARN  [remote_server] (remote server) error reading message on stdin, dropping connection.
2026-05-03T22:06:14+08:00 INFO  [remote_server] (remote server) accepting new connections
2026-05-03T22:06:14+08:00 INFO  [remote_server] (remote server) accepted new connections
2026-05-03T22:06:14+08:00 INFO  [remote::remote_client] Successfully reconnected
2026-05-03T22:06:14+08:00 INFO  [remote::remote_client] Successfully reconnected
2026-05-03T22:06:18+08:00 INFO  [zed::zed] Using GPU: GpuSpecs { is_software_emulated: false, device_name: "Intel(R) UHD Graphics", driver_name: "Intel Corporation", driver_info: "32.0.101.6790" }
2026-05-03T22:06:18+08:00 WARN  [crashes] Failed to send data to crash server Io(Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." })
2026-05-03T22:06:18+08:00 ERROR [crates/agent_ui/src/agent_panel.rs:155] Error("unknown variant `Codex`, expected one of `NativeAgent`, `TextThread`, `native_agent`, `Custom`, `custom`", line: 1, column: 39)
2026-05-03T22:06:18+08:00 INFO  [project::prettier_store] Initializing default prettier with plugins {}
2026-05-03T22:06:19+08:00 INFO  [lsp] starting language server process. binary path: "D:\\Program Files\\nodejs\\node.exe", working directory: "C:\\Users\\Cheng\\AppData\\Roaming\\Zed", args: ["C:\\Users\\Cheng\\AppData\\Local\\Zed\\languages\\json-language-server\\node_modules/vscode-langservers-extracted/bin/vscode-json-language-server", "--stdio"]
2026-05-03T22:06:19+08:00 INFO  [project::prettier_store] Installing default prettier and plugins: [("prettier", "3.8.3")]
2026-05-03T22:06:20+08:00 INFO  [project::prettier_store] Initialized default prettier with plugins: {}
2026-05-03T22:08:04+08:00 WARN  [remote::remote_client] remote heartbeat: connection activity channel has been dropped. stopping.
2026-05-03T22:08:04+08:00 ERROR [remote::remote_client] proxy process terminated unexpectedly: -1
2026-05-03T22:08:04+08:00 INFO  [remote::remote_client] setting state to 'reconnecting'
2026-05-03T22:08:04+08:00 INFO  [remote::remote_client] Trying to reconnect to remote server... Attempt 1
2026-05-03T22:08:04+08:00 WARN  [remote::remote_client] remote heartbeat: connection activity channel has been dropped. stopping.
2026-05-03T22:08:04+08:00 ERROR [remote::remote_client] proxy process terminated unexpectedly: -1
2026-05-03T22:08:04+08:00 INFO  [remote::remote_client] setting state to 'reconnecting'
2026-05-03T22:08:04+08:00 INFO  [remote::remote_client] Trying to reconnect to remote server... Attempt 1
2026-05-03T22:08:14+08:00 INFO  [remote::transport::ssh] Remote is windows: false
2026-05-03T22:08:16+08:00 INFO  [remote::transport::ssh] Remote shell discovered: /bin/bash
2026-05-03T22:08:17+08:00 INFO  [remote::transport::ssh] Remote platform discovered: RemotePlatform { os: Linux, arch: X86_64 }
2026-05-03T22:08:21+08:00 WARN  [remote_server] (remote server) stdin read failed: unexpected end of file
2026-05-03T22:08:21+08:00 WARN  [remote_server] (remote server) error reading message on stdin, dropping connection.
2026-05-03T22:08:21+08:00 INFO  [remote_server] (remote server) accepting new connections
2026-05-03T22:08:21+08:00 INFO  [remote_server] (remote server) accepted new connections
2026-05-03T22:08:21+08:00 INFO  [remote::remote_client] Successfully reconnected
2026-05-03T22:08:24+08:00 ERROR [remote::remote_client] Reconnect attempt 1 failed: Timed out resyncing remote client. Starting new attempt...
2026-05-03T22:08:24+08:00 INFO  [remote::remote_client] setting state to 'reconnecting'
2026-05-03T22:08:24+08:00 INFO  [remote::remote_client] Trying to reconnect to remote server... Attempt 2
2026-05-03T22:08:31+08:00 INFO  [remote::transport::ssh] Remote is windows: false
2026-05-03T22:08:32+08:00 INFO  [remote::transport::ssh] Remote shell discovered: /bin/bash
2026-05-03T22:08:34+08:00 INFO  [remote::transport::ssh] Remote platform discovered: RemotePlatform { os: Linux, arch: X86_64 }
2026-05-03T22:08:37+08:00 WARN  [remote_server] (remote server) error reading message on stdin, dropping connection.
2026-05-03T22:08:37+08:00 INFO  [remote_server] (remote server) accepting new connections
2026-05-03T22:08:37+08:00 INFO  [remote_server] (remote server) accepted new connections
2026-05-03T22:08:38+08:00 INFO  [remote::remote_client] Successfully reconnected
2026-05-03T22:08:50+08:00 INFO  [zed::zed] Using GPU: GpuSpecs { is_software_emulated: false, device_name: "Intel(R) UHD Graphics", driver_name: "Intel Corporation", driver_info: "32.0.101.6790" }
2026-05-03T22:08:50+08:00 WARN  [crashes] Failed to send data to crash server Io(Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." })
2026-05-03T22:08:50+08:00 ERROR [crates/agent_ui/src/agent_panel.rs:155] Error("unknown variant `Codex`, expected one of `NativeAgent`, `TextThread`, `native_agent`, `Custom`, `custom`", line: 1, column: 39)
2026-05-03T22:08:50+08:00 INFO  [project::prettier_store] Initializing default prettier with plugins {}
2026-05-03T22:08:50+08:00 INFO  [lsp] starting language server process. binary path: "D:\\Program Files\\nodejs\\node.exe", working directory: "C:\\Users\\Cheng\\AppData\\Roaming\\Zed", args: ["C:\\Users\\Cheng\\AppData\\Local\\Zed\\languages\\json-language-server\\node_modules/vscode-langservers-extracted/bin/vscode-json-language-server", "--stdio"]

```

</details>

### Relevant Zed settings

<details><summary>settings.json</summary>

```json

```

</details>

### Relevant Keymap

<details><summary>keymap.json</summary>

```json

```

</details>

### If you are using WSL on Windows, what flavor of Linux are you using?

None

**Comment by yeskunall** [comment]
See this https://github.com/zed-industries/zed/issues/50237#issuecomment-4011591079. This was reported in https://github.com/zed-industries/zed/issues/50237, but the PR that closed it fixed it only for WSL, not for remote connections via SSH.

@brucegorman -- that’s because you’re remoting into the machine and then trying to open the settings file:

```
INFO  [remote::transport::ssh] Remote shell discovered: /bin/bash
INFO  [remote::transport::ssh] Remote platform discovered: RemotePlatform { os: Linux, arch: X86_64 }
```

**Comment by lingyaochu** [comment]
The case is, directly opening `settings.json` is unavailable in a remote session. But if you first open the settings panel via `zed: open settings`, and then click the `Edit in settings.json` button, it will open the file. This can serve as a workaround before this issue gets solved.

**Comment by xingxing21** [comment]
Hello @yeskunall 
I’d like to work on this issue if it’s available.
Could you please assign it to me? 😉 
Best regards,
Xing

**Comment by xingxing21** [comment]
@yeskunall 

### Root Cause
The bug is in open_settings_file in crates/zed/src/zed.rs:2352. When triggered from an SSH remote workspace, the function uses with_local_or_wsl_workspace which — for SSH remote — creates a new local window and calls the provided callback there. The problem has two parts:

1. Problematic double-nesting of with_local_or_wsl_workspace:

open_settings_file's callback calls create_and_open_local_file (workspace.rs:9988), which itself calls with_local_or_wsl_workspace a second time internally (at line 10004). So the chain is:


open_settings_file
  → with_local_or_wsl_workspace (SSH: creates new local window, runs callback)
    → callback calls create_and_open_local_file
      → create_and_open_local_file calls with_local_or_wsl_workspace AGAIN (now local branch, OK)
The outer callback returns two tasks (worktree_creation_task, settings_open_task) that are then awaited from the SSH workspace's async context — a cross-window task dependency that is fragile.

2. Missing window.activate_window():

create_and_open_local_file never activates the window. So when a new local window IS successfully created and the file IS opened, the window appears behind the SSH workspace window with no indication that anything happened.

### Why the Settings Panel Workaround Works
open_current_settings_file in [settings_ui.rs:3453] calls open_user_settings_in_workspace (settings_ui.rs:3882), which:

- Is itself the callback passed to with_local_or_wsl_workspace (no double-nesting)
- Spawns a self-contained task that runs independently via .detach()
- Explicitly calls window.activate_window() at the end (line 3920)
